OpenRA.Mods.Common/Widgets/Logic/Ingame/ArmyTooltipLogic.cs

```diff
@@ -0,0 +1,70 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Linq;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Widgets;
+
+namespace OpenRA.Mods.Common.Widgets.Logic
+{
+	public class ArmyTooltipLogic : ChromeLogic
+	{
+		[ObjectCreator.UseCtor]
+		public ArmyTooltipLogic(Widget widget, TooltipContainerWidget tooltipContainer, Func<ActorInfo> getTooltipUnit)
+		{
+			widget.IsVisible = () => getTooltipUnit() != null;
+			var nameLabel = widget.Get<LabelWidget>("NAME");
+
+			var descLabel = widget.Get<LabelWidget>("DESC");
+
+			var font = Game.Renderer.Fonts[nameLabel.Font];
+			var descFont = Game.Renderer.Fonts[descLabel.Font];
+
+			ActorInfo lastActor = null;
+			var descLabelPadding = descLabel.Bounds.Height;
+
+			tooltipContainer.BeforeRender = () =>
+			{
+				var actor = getTooltipUnit();
+				if (actor == null)
+					return;
+
+				if (actor == lastActor)
+					return;
+
+				var tooltip = actor.TraitInfos<TooltipInfo>().FirstOrDefault(info => info.EnabledByDefault);
+				var name = tooltip != null ? tooltip.Name : actor.Name;
+				var buildable = actor.TraitInfo<BuildableInfo>();
```

Comment:
If we go with my suggestion for the overhauled `Unit` class (now that I think about it, we probably want to rename that to something like `ArmyUnit`) we could change `getTooltipUnit` to return it, have it include `TooltipInfo[]`, and remove these info queries.